File: src/main/java/com/educative/grokking/exercises/GroupAnagrams.java
package com.educative.grokking.exercises;

import java.util.*;

// https://leetcode.com/problems/group-anagrams/
public class GroupAnagrams {
    public static List<List<String>> groupAnagrams(String[] strs) {
        var map = new HashMap<ArrayList<Integer>, List<String>>();

        for (var s : strs) {
            var counts = new ArrayList<Integer>();
            for (int i = 0; i < 26; i++) {
                counts.add(0);
            }

            for (var c : s.toCharArray()) {
                counts.set(c - 'a', counts.get(c - 'a') + 1);
            }


            map.computeIfAbsent(counts, k -> new ArrayList<>()).add(s);
        }

        var result = new ArrayList<List<String>>();
        for (var e : map.entrySet()) {
            result.add(e.getValue());
        }
        return result;
    }
}

File: src/main/kotlin/neetcode/GroupAnagrams.kt
package neetcode

// https://leetcode.com/problems/group-anagrams/
class GroupAnagrams {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val arrayOfCount26 = strs.map { str ->
            val count26 = IntArray(26) { 0 }
            str.toCharArray().map { c ->
                count26[c - 'a'] += 1
            }
            str to count26.joinToString()
        }

        val groupByCounts26 = HashMap<String, ArrayList<String>>()
        arrayOfCount26.forEach { (str, count26) ->
            groupByCounts26
                .computeIfAbsent(count26) { ArrayList<String>() }
                .add(str)
        }

        return groupByCounts26.values.toList()

    }
}